crates/transpiler/src/passes/sabre/dag.rs

```diff
@@ -149,28 +184,54 @@ impl SabreDAG {
                 panic!("op nodes should always be of type `Operation`");
             };
             let kind = InteractionKind::from_op(&inst.op, dag.get_qargs(inst.qubits))?;
-            let sabre_node = sabre.add_node(SabreNode::new(dag_node, kind));
-            for edge in dag.dag().edges_directed(dag_node, Direction::Incoming) {
-                if let Some(parent) = wire_pos.insert(*edge.weight(), sabre_node) {
-                    sabre.add_edge(parent, sabre_node, ());
-                }
-            }
-            if sabre
-                .edges_directed(sabre_node, Direction::Incoming)
-                .next()
-                .is_none()
-            {
-                match &sabre[sabre_node].kind {
+            match predecessors(dag_node, &wire_pos) {
+                Predecessors::AllUnmapped => match kind {
                     InteractionKind::Synchronize => {
+                        initial.push(dag_node);
+                    }
+                    kind => {
```

Comment:
There a so many weird match syntax things I don't know about. I really didn't know this worked until just now.